File: 0776_split_bst/Solution.cs
// LeetCode 0776 - Split BST
// https://leetcode.com/problems/split-bst/

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
public class Solution {
    public TreeNode[] SplitBST(TreeNode root, int target) {
        if (root == null) return new TreeNode[] { null, null };
        if (root.val <= target) {
            var parts = SplitBST(root.right, target);
            root.right = parts[0];
            return new TreeNode[] { root, parts[1] };
        }
        var leftParts = SplitBST(root.left, target);
        root.left = leftParts[1];
        return new TreeNode[] { leftParts[0], root };
    }
}

File: 0776_split_bst/Solution.kt
// LeetCode 0776 - Split BST
// https://leetcode.com/problems/split-bst/

class TreeNode(var `val`: Int = 0) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class Solution {
    fun splitBST(root: TreeNode?, target: Int): Array<TreeNode?> {
        if (root == null) return arrayOf(null, null)
        if (root.`val` <= target) {
            val parts = splitBST(root.right, target)
            root.right = parts[0]
            return arrayOf(root, parts[1])
        }
        val leftParts = splitBST(root.left, target)
        root.left = leftParts[1]
        return arrayOf(leftParts[0], root)
    }
}
